cfgov/archival/apps.py

```diff
@@ -0,0 +1,13 @@
+from django.apps import AppConfig
+from django.conf import settings
+
+import fs
+
+
+class ArchivalConfig(AppConfig):
+    name = "archival"
+    label = "archival"
+    filesystem_name = getattr(settings, "ARCHIVE_FILESYSTEM", None)
+    filesystem = (
+        fs.open_fs(filesystem_name) if filesystem_name is not None else None
+    )
```

Comment:
Is there some way to defer this call so that it doesn't happen on import time, but rather later, perhaps when the filesystem is first accessed? As written it seems like the {file, S3, etc} would be accessed whenever this file is imported by Python.

Reply:
I'll check — this is how [django-bakery uses `fs` to configure its storage](https://github.com/palewire/django-bakery/blob/main/bakery/apps.py#L11-L12), so that's what I'm doing here. 

> As written it seems like the {file, S3, etc} would be accessed whenever this file is imported by Python.

My understanding from [fs's documentation](https://docs.pyfilesystem.org/en/latest/guide.html#opening-filesystems) is that this constructs the object that accesses the filesystem, but doesn't perform access operations until those are called. This is [reflected in the S3FS constructor itself too](https://github.com/PyFilesystem/s3fs/blob/master/fs_s3fs/_s3fs.py#L265-L306). S3 isn't accessed when it's initialized, the accessor object is just initialized. You can confirm this in a Python shell:

```python
>>> from fs import open_fs
>>> s3 = open_fs('s3://some:creds@wills-non-existent-bucket')
>>> s3
S3FS('wills-non-existent-bucket')
>>> s3.listdir("/")
Traceback (most recent call last):
  …
botocore.exceptions.ClientError: An error occurred (InvalidAccessKeyId) when calling the ListObjects operation: The AWS Access Key Id you provided does not exist in our records.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  …
fs.errors.PermissionDenied: The AWS Access Key Id you provided does not exist in our records.
```

The reason django-bakery does this I'm guessing, and I'm doing it here, is that this way all access to the selected filesystem goes through a single object.

Reply:
Thanks! I'd wondered if this could be the case so nice to know it is.